### PR TITLE
HSEARCH-3787 Support hard timeout on Hibernate ORM entity loading

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3787
~~https://hibernate.atlassian.net/browse/HSEARCH-3774~~ removed from this PR's scope

### CODE REVIEW

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingBaseIT.java [hunk 1]
```diff
@@ -62,6 +63,22 @@ public void simple() {
...
+
+		persistThatManyEntities( entityCount );
+
+		testLoadingThatManyEntities(
+				session -> { }, // No particular session setup
+				o -> { }, // No particular loading option
+				entityCount,
+				// Only one entity type means only one statement should be executed, even if there are multiple hits
+				c -> c.assertStatementExecutionCount().isEqualTo( 1 ),
+				1, TimeUnit.SECONDS
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingBaseIT.java:78]: You can't just put a timeout and not make any assertion related to that timeout. The test, as it is, is just a smoke test.

You need to test at least these cases:

* `failAfter` with a very large timeout: expect everything to work fine.
* `failAfter` with a small-ish timeout (say, 1ms) and very slow loading: expect an exception to be thrown.
* `failAfter` with a large timeout (say, 500ms), very slow index response (say, 499ms) and normal loading: expect an exception to be thrown.
* And maybe more. Have a look at your code and see if some obvious edge cases are not covered by the above?

To create the right conditions for your tests (slow index response time, slow loading), you have two options.

The cleanest approach would be to stub the `TimingSource`. However, it might not be very easy, and if done incorrectly you may end up not testing what you want.

The admittedly more dirty, but more straightforward approach is to voluntarily slow down your components:

* To simulate slow index response time, just pass a custom behavior to `org.hibernate.search.util.impl.integrationtest.common.rule.BackendMock#expectSearchObjects(java.util.Collection<java.lang.String>, java.util.function.Consumer<org.hibernate.search.util.impl.integrationtest.common.stub.backend.search.StubSearchWork.Builder>, org.hibernate.search.util.impl.integrationtest.common.rule.StubSearchWorkBehavior<org.hibernate.search.engine.backend.common.DocumentReference>)`, and in that behavior call `Thread.sleep()`.
* To simulate slow loading, you can add a `SessionEventListener` through `SessionBuilderImpl#eventListeners`, and in `SessionEventListener#jdbcPrepareStatementEnd()` call `Thread.sleep()`.

File: util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java [hunk 2]
```diff
@@ -122,4 +123,16 @@ public String toString() {
...
+	static Integer getLoadingTimeout(StubSearchWork work) {
+		Long timeout = work.getFailAfterTimeout();
+		TimeUnit timeUnit = work.getFailAfterTimeUnit();
+		if ( timeout == null || timeUnit == null ) {
+			return null;
+		}
+
+		long nanos = timeUnit.toNanos( timeout );
+		int millis = Math.toIntExact( nanos / 1000000 );
+		return ( nanos % 1000000 == 0 ) ? millis : millis + 1;
```
Comment (yrodiere) [inline util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java:135]: I believe this code was copy/pasted from the Lucene backend? Maybe it would be a good idea to move it to some shared util class.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java [hunk 3]
```diff
@@ -81,13 +81,17 @@
 		Integer took = TOOK_ACCESSOR.get( responseBody ).get();
 		Boolean timedOut = TIMED_OUT_ACCESSOR.get( responseBody ).get();
 
+		Integer remainingTimeToHardTimeout = ( hardTimeoutInMilliseconds == null ) ? null :
+				Math.toIntExact( hardTimeoutInMilliseconds ) - took;
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java]: That's hardly a good way to compute the remaining time. `took` is the execution time on the server side. It ignores network transfer time completely.

The "right" way to handle these things would be to keep track of when the query started executing on the client side. Then here you would be able to know exactly how much time elapsed. That's exactly what we do in the Lucene backend, and I believe that would be the only correct solution here too.

On a side note, this solution would also allow us to solve https://hibernate.atlassian.net/browse/HSEARCH-3774 (in another PR, obviously).

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java [hunk 4]
```diff
@@ -47,7 +47,8 @@
 				referencesByDelegate.entrySet() ) {
 			HibernateOrmComposableEntityLoader<? extends T> delegate = entry.getKey();
 			List<EntityReference> referencesForDelegate = entry.getValue();
-			delegate.loadBlocking( referencesForDelegate, objectsByReference );
+			// FIXME Use a deadline instead of a timeout
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java]: Yep. Pass the `TimeoutManager`, for instance.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java [hunk 5]
```diff
@@ -113,13 +114,16 @@ public void loadBlocking(List<EntityReference> references, Map<? super EntityRef
...
-	private Query<? extends E> createQuery(int fetchSize) {
+	private Query<? extends E> createQuery(int fetchSize, Integer timeout) {
 		Query<? extends E> query = (Query<? extends E>) HibernateOrmUtils.createQueryForLoadByUniqueProperty(
 				session, entityPersister, documentIdSourcePropertyName,
 				DOCUMENT_ID_SOURCE_PROPERTY_PARAMETER_NAME
 		);
 
 		query.setFetchSize( fetchSize );
+		if ( timeout != null ) {
+			query.setTimeout( HibernateOrmComposableEntityLoader.getTimeoutInSeconds( timeout ) );
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java:125]: Rounding up the timeout to the next second is definitely not acceptable for search queries. People need sub-second response times for the overwhelming majority of use cases.

Use `query.setHint( QueryHints.SPEC_HINT_TIMEOUT, <value in milliseconds> )` instead.

File: engine/src/main/java/org/hibernate/search/engine/common/impl/DelegatingBuildContext.java [hunk 6]
```diff
@@ -44,4 +46,12 @@ public ThreadPoolProvider threadPoolProvider() {
 	public FailureHandler failureHandler() {
 		return delegate.getFailureHandler();
 	}
+
+	public EngineThreads getEngineThreads() {
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/common/impl/DelegatingBuildContext.java:50]: I think you mean `engineThreads()`? :)

File: engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationBuilderImpl.java [hunk 7]
```diff
@@ -194,11 +199,14 @@ public SearchIntegrationPartialBuildState prepareBuild() {
 
 			threadProviderHolder = THREAD_PROVIDER.getAndTransform( propertySource, beanResolver::resolve );
 			ThreadPoolProviderImpl threadPoolProvider = new ThreadPoolProviderImpl( threadProviderHolder );
+			engineThreads = new EngineThreads( threadPoolProvider );
+			timingSource = new DefaultTimingSource( engineThreads );
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationBuilderImpl.java:203]: There's a `SuppressingCloser` below; you should probably close these with that closer?

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/impl/LuceneBackendImpl.java [hunk 8]
```diff
@@ -19,14 +19,14 @@
 import org.hibernate.search.backend.lucene.multitenancy.impl.MultiTenancyStrategy;
 import org.hibernate.search.backend.lucene.orchestration.impl.LuceneSyncWorkOrchestratorImpl;
 import org.hibernate.search.backend.lucene.resources.impl.BackendThreads;
-import org.hibernate.search.backend.lucene.common.timing.impl.TimingSource;
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/impl/LuceneBackendImpl.java:22]: You should probably merge the commit `HSEARCH-3787 Remove lucene-level TimingSource ` with the previous one. Otherwise you're missing out on git's ability to track renamed files and to provide a commit history for a particular file even beyond its renaming.

File: engine/src/main/java/org/hibernate/search/engine/common/timing/impl/TimingSource.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.common.timing.impl;
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/common/timing/impl/TimingSource.java:7]: If you're going to expose `TimingSource` to other modules (e.g. the backends), you will need to put this interface in an `spi` package.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/timeout/impl/LuceneTimeoutManager.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		timingSource.ensureInitialized();
+	}
+
+	public Counter createCounter() {
+		return new LuceneCounterAdapter( timingSource );
+	}
+
+	@Override
+	protected void onTimedOut() {
+		if ( hasHardTimeout() ) {
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/timeout/impl/LuceneTimeoutManager.java:67]: I think this should be moved to the superclass, too. Be it only because it's simpler that way, or because we also support "soft" timeouts with the Elasticsearch backend. You could get rid of a few class members in the Elasticsearch backend if it was able to query `hasHardTimeout` directly.

File: mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/spi/PojoMappingDelegate.java [hunk 11]
```diff
@@ -30,6 +31,8 @@
 
 	FailureHandler failureHandler();
 
+	TimingSource timingSource();
```
Comment (yrodiere) [inline mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/spi/PojoMappingDelegate.java:34]: Wouldn't it be simpler to just expose the `TimeoutManager` to the mapper, and not the `TimingSource`? I thought that was the reason you moved the `TimeoutManager` to the engine... I'm a bit confused.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java [hunk 12]
```diff
@@ -23,13 +25,21 @@
...
-	HibernateOrmByTypeEntityLoader(Map<String, HibernateOrmComposableEntityLoader<? extends T>> delegatesByEntityName) {
+	HibernateOrmByTypeEntityLoader(Map<String, HibernateOrmComposableEntityLoader<? extends T>> delegatesByEntityName,
+			TimingSource timingSource) {
 		this.delegatesByEntityName = delegatesByEntityName;
+		this.timingSource = timingSource;
 	}
 
 	@Override
 	public List<T> loadBlocking(List<EntityReference> references, Integer timeout) {
+		TimeoutManager timeoutManager = null;
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java:38]: Here, you could just pass the `TimeoutManager` to this method instead of the `Integer timeout`. I expect this will make a lot of code simpler and safer.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java [hunk 13]
```diff
@@ -81,8 +81,8 @@
 		Integer took = TOOK_ACCESSOR.get( responseBody ).get();
 		Boolean timedOut = TIMED_OUT_ACCESSOR.get( responseBody ).get();
 
-		Integer remainingTimeToHardTimeout = ( hardTimeoutInMilliseconds == null ) ? null :
-				Math.toIntExact( hardTimeoutInMilliseconds ) - took;
+		Long remainingTimeToHardTimeout = ( hardTimeoutInMilliseconds == null ) ? null :
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java:84]: Yes, using `long` all the way is safer.

File: integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/testsupport/stub/MapperEasyMockUtils.java [hunk 14]
```diff
@@ -91,4 +101,22 @@ private MapperEasyMockUtils() {
...
+				EasyMockUtils.collectionAnyOrderMatcher( new ArrayList<>( context.loadingMap.keySet() ) ), isNull() ) )
+				.andAnswer( () -> ( (List<R>) EasyMock.getCurrentArguments()[0] ).stream()
+						.map( context.loadingMap::get )
+						.collect( Collectors.toList() ) );
+	}
+
+	private static <R, E> void loadWithTimeout(EntityLoader<R, E> objectLoaderMock,
+			HitMappingDefinitionContext<R, E> context) {
+		expect( objectLoaderMock.loadBlocking(
+				EasyMockUtils.collectionAnyOrderMatcher( new ArrayList<>( context.loadingMap.keySet() ) ), notNull() ) )
```
Comment (yrodiere) [inline integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/testsupport/stub/MapperEasyMockUtils.java:117]: Right, good idea. You could even make this more powerful and pass a `Capture<Long>`, so that in your tests, you can check that the number is within a given range.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/SearchQueryBaseIT.java [hunk 15]
```diff
@@ -178,6 +180,34 @@ public void target_byClass_multipleTypes() {
 		} );
 	}
 
+	@Test
+	public void target_byClass_multipleTypes_entityLoadingTimeout() {
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/SearchQueryBaseIT.java:184]: See my comment on your other commit adding mapper-level tests: you need to simulate slow loading, at least.

Also, I'm sure this test currently passes (from time to time) only because your implementation rounds the timeout value to the nearest second.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingMultipleTypesIT.java [hunk 16]
```diff
@@ -227,6 +231,52 @@ public void mixedHierarchies() {
 		);
 	}
 
+	@Test
+	public void mixedHierarchies_entityLoadingTimeout() {
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingMultipleTypesIT.java:235]: Same here: you need to simulate slow loading, at least.

Also, I'm sure this test currently passes (from time to time) only because your implementation rounds the timeout value to the nearest second.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java (hunk 3 above)
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/Elasticsearch7SearchResultExtractor.java]: Agreed. I assigned the issue to myself.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java (hunk 5 above)
Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java:125]: Right! I forgot we have this option

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/timeout/impl/LuceneTimeoutManager.java (hunk 10 above)
Comment (fax4ever) [reply backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/timeout/impl/LuceneTimeoutManager.java:67]: Sure. Originally I didn't think that Elasicsearch could use it.

File: integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/testsupport/stub/MapperEasyMockUtils.java (hunk 14 above)
Comment (fax4ever) [reply integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/testsupport/stub/MapperEasyMockUtils.java:117]: yep makes sense to use a spy for that

File: util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java (hunk 2 above)
Comment (fax4ever) [reply util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java:135]: In Lucene backend we're using

``` java
timeoutUnit.toMillis( timeoutValue )
```
Probably it would be better doing the same here.

Comment (fax4ever) [reply util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java:135]: Or maybe not, since the method handles null inputs and always perform a round up of the values.
Moving to an util class...

``` java
@Test
public void test() {
	long l = TimeUnit.NANOSECONDS.toMillis( 1L );
	Assertions.assertThat( l ).isEqualTo( 0 );

	Long aLong = ConvertUtils.toMilliseconds( 1L, TimeUnit.NANOSECONDS );
	Assertions.assertThat( aLong ).isEqualTo( 1L );
}
```

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java (hunk 5 above)
Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java:125]: Anyway I'm not sure that with `QueryHints.SPEC_HINT_TIMEOUT` we can have more precision. See `AbstractProducedQuery#setHint`, in particular the if branch when `SPEC_HINT_TIMEOUT.equals( hintName )`.

Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java:125]: Probably it is better to use `HINT_TIMEOUT`.

File: engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationBuilderImpl.java (hunk 7 above)
Comment (fax4ever) [reply engine/src/main/java/org/hibernate/search/engine/common/impl/SearchIntegrationBuilderImpl.java:203]: Yes the idea was to have the variables declared in the outer block for that... but I forget to add the code lines to close them on error on startup...

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java (hunk 5 above)
Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmNonEntityIdPropertyEntityLoader.java:125]: > Anyway I'm not sure that with QueryHints.SPEC_HINT_TIMEOUT we can have more precision. See AbstractProducedQuery#setHint, in particular the if branch when SPEC_HINT_TIMEOUT.equals( hintName ).

Well, that's underwhelming... Looks like JDBC only allows second precision. I guess at least we'll be ready for a potential future version of the spec, if they ever have a look at that...

What's more annoying is how are we going to test this... We need a query that takes significantly more than one second to execute, or we need to mock the `TimingSource` (which has its own disadvantage, as I mentioned over the chat). Let's discuss this on Zulip if you need to.

File: mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/spi/PojoMappingDelegate.java (hunk 11 above)
Comment (fax4ever) [reply mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/spi/PojoMappingDelegate.java:34]: In the current solution the engine can provide the timing source, this instance is shared with all the modules that need it. On the other hand the timeout manager is created each time a module needs it by the module itself. So that the module can control its lifecycle. Another benefit is the fact that a client module can choose its own subclass implementation, e.g.: `LuceneTimeoutManager`.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingBaseIT.java (hunk 1 above)
Comment (fax4ever) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/search/loading/SearchQueryEntityLoadingBaseIT.java]: I added extensive (I think) for the first and second cases. About the last case you mentioned. IMO this would test the time capability of the stub backend to track time and to pass the right difference to the mapper. I don't thing it would make much sense.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java [hunk 17]
```diff
@@ -47,7 +48,12 @@
 				referencesByDelegate.entrySet() ) {
 			HibernateOrmComposableEntityLoader<? extends T> delegate = entry.getKey();
 			List<EntityReference> referencesForDelegate = entry.getValue();
-			delegate.loadBlocking( referencesForDelegate, objectsByReference );
+
+			delegate.loadBlocking( referencesForDelegate, objectsByReference, timeoutManager );
+		}
+
+		if ( timeoutManager != null ) {
+			timeoutManager.stop();
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java:56]: I believe this should be called by whoever called `start`, not by this class.

File: util/internal/integrationtest/mapper/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/mapper/orm/TimeoutLoadingListener.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+package org.hibernate.search.util.impl.integrationtest.mapper.orm;
+
+import org.hibernate.BaseSessionEventListener;
+import org.hibernate.QueryTimeoutException;
+import org.hibernate.Session;
+import org.hibernate.engine.spi.SessionImplementor;
+
+public class TimeoutLoadingListener extends BaseSessionEventListener {
+
+	public static void registerSlowerLoadingListener(Session session) {
```
Comment (yrodiere) [inline util/internal/integrationtest/mapper/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/mapper/orm/TimeoutLoadingListener.java:16]: Maybe `registerTimingOutLoadingListener`?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/timeout/impl/ElasticsearchTimeoutManager.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	@Override
+	protected void onTimedOut() {
+		if ( hasHardTimeout() ) {
+			throw log.clientSideTimedOut( Duration.ofNanos( timeoutUnit.toNanos( timeoutValue ) ), query.toString() );
+		}
+	}
+
+	public boolean exceptionOnTimeout() {
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/timeout/impl/ElasticsearchTimeoutManager.java:57]: There's already a method that does that: `hasHardTimeout`.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java (hunk 17 above)
Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/HibernateOrmByTypeEntityLoader.java:56]: Good catch. Right, the timeout lifecycle is handled by the backends.

**Comment by fax4ever** [comment]
Thanks for the review @yrodiere.

As you pointed out, I removed the HSEARCH-3774 from the scope of this pull request and did the three changes you suggested above. I'm going to create a new pull request to handle HSEARCH-3774  properly.
Meanwhile, is there anything else we can improve here?

**Comment by fax4ever** [comment]
Here is a build with all supported evns: https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2333/21/

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java [hunk 20]
```diff
@@ -222,6 +234,7 @@ private Integer defaultedLimit(Integer limit, Integer offset) {
 	}
 
 	private JsonObject doExplain(ElasticsearchSearchIndexContext index, Object id) {
+		timeoutManager.start();
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java:237]: I just noticed this... you're not supposed to use the timeout manager during `explain()` (cf. documentation). At least we made sure not to use it in the Lucene backend.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java [hunk 21]
```diff
@@ -141,23 +144,32 @@ public long fetchTotalHitCount() {
...
 		NonBulkableWork<Long> work = builder.build();
-		return Futures.unwrappedExceptionJoin( queryOrchestrator.submit( work ) );
+		Long result = Futures.unwrappedExceptionJoin( queryOrchestrator.submit( work ) );
+		timeoutManager.stop();
+		return result;
 	}
 
 	@Override
 	public ElasticsearchSearchScroll<H> scroll(int chunkSize) {
+		timeoutManager.start();
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java:159]: I just noticed this... you're not supposed to use the timeout manager during the initial call to `scroll()`, since we don't send a request to Elasticsearch.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java (hunk 20 above)
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java:237]: I agree, I'm removing it...

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java (hunk 21 above)
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java:159]: Actually with the initial call of scroll() on Elasticsearch we got the first result page of next() you know. So, IMO makes sense to keep it.

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchQueryImpl.java]: Done and more important this remove discover an inner bug solved by this commit https://github.com/hibernate/hibernate-search/pull/2333/commits/6e95d93879d1d99a0b53e26bfe2c1b7154bd608d.

**Comment by fax4ever** [comment]
Address the comments, thanks. When a full CI build passes I'll merge it.
